File: go/luhn/luhn.go
package luhn

import (
	"unicode"
)

// Valid calculates if a given number
// is valid per the Luhn formula.
func Valid(input string) bool {
	sum := 0
	count := 0

	for _, char := range input {
		if unicode.IsSpace(char) {
			continue
		}

		if !unicode.IsDigit(char) {
			return false
		}

		digit := int(char - '0')
		if count%2 == len(input)%2 {
			digit *= 2
			if digit > 9 {
				digit -= 9
			}
		}
		sum += digit
		count++
	}

	return count > 1 && sum%10 == 0
}

File: python/luhn/luhn.py
class Luhn:
    def __init__(self, card):
        self.card = card

    def valid(self):
        total = 0
        count = 0

        for c in self.card[::-1]:
            if c == ' ':
                continue

            if not str.isdigit(c):
                return False

            digit = int(c)
            digit = 2*digit if count % 2 > 0 else digit
            digit = digit-9 if digit > 9 else digit

            total += digit
            count += 1

        return count > 1 and total % 10 == 0

